Main.hs

```diff
@@ -11,28 +11,32 @@ import qualified Data.List as L
 import Text.Printf (printf)
 
 multByIndex :: [Int] -> [Int]
-multByIndex = undefined
+multByIndex xs = zipWith (*) xs [0..]
 
 powerByIndex :: [Int] -> [Int]
-powerByIndex = undefined
+powerByIndex xs = zipWith (^) xs [0..]
+
 
 productOfDifference :: [Int] -> Int
-productOfDifference = undefined
+productOfDifference xs
+    | null xs = error "The list is empty"
+    | xs == [head xs] = error "The list is too short"
+    | otherwise = foldl (*) 1 (zipWith (-) xs (tail xs))
 
-isSorted :: [Int] -> Bool 
-isSorted xs = undefined
+isSorted :: [Int] -> Bool
+isSorted xs = and (zipWith (<=) xs (tail xs))
 
-countElement :: Int -> [Int] -> Int 
-countElement = undefined
+countElement :: Int -> [Int] -> Int
+countElement x = length . filter (x==)
 
-dotProduct :: [Int] -> [Int] -> Int 
-dotProduct = undefined
+dotProduct :: [Int] -> [Int] -> Int
+dotProduct xs ys = foldl (+) 0 (zipWith (*) xs ys)
 
 applyAll :: [a -> b] -> a -> [b]
-applyAll = undefined
+applyAll fs x = map (\f -> f x) fs
```

Comment:
You can use operator sections to avoid a lambda-function. Could you figure out how?